src/GithubPlugin.fs

```diff
@@ -18,21 +20,28 @@ type GithubPlugin() =
 
     let runApiSearch = Gh.runSearchCached >> RunApiSearch
 
-    let parseQuery = function
-        | [ "repos"; search ]                     -> runApiSearch (FindRepos search)
-        | [ "users"; search ]                     -> runApiSearch (FindUsers search)
-        | [ "issues"; UserRepoFormat search ]     -> runApiSearch (FindIssues search)
-        | [ "pr";     UserRepoFormat search ]     -> runApiSearch (FindPRs search)
-        | [ "pull";   UserRepoFormat search ]     -> runApiSearch (FindPRs search)
-        | [ "repo";   UserRepoFormat search ]     -> runApiSearch (FindRepo search)
-        | [ UserRepoFormat search           ]     -> runApiSearch (FindRepo search)
-        | [ UserRepoFormat search; "issues" ]     -> runApiSearch (FindIssues search)
-        | [ UserRepoFormat search; "pr"     ]     -> runApiSearch (FindPRs search)
-        | [ UserRepoFormat search; "pull"   ]     -> runApiSearch (FindPRs search)
-        | [ UserRepoFormat (u,r); IssueFormat i ] -> runApiSearch (FindIssue (u,r,i))
-        | [ UserReposFormat user ]                -> runApiSearch (FindUserRepos user)
-        | [ search ]                              -> SuggestQuery (SearchRepos search)
-        | _                                       -> SuggestQuery DefaultSuggestion
+    let (|SpecificSearchFormat|_|) keyword =
+        match keyword with
+        | "issues" -> Some(FindIssues)
+        | "pr"
+        | "pull" -> Some(FindPRs)
+        | "repo" -> Some(FindRepo)
+        | _ -> None
+
+    let (|GeneralSearchFormat|_|) keyword =
+        match keyword with
+        | "repos" -> Some(FindRepos)
+        | "users" -> Some(FindUsers)
+        | _ -> None
+
+    let parseQuery (searchTerm: Query) =
+        match (searchTerm.FirstSearch, searchTerm.SecondToEndSearch) with
+        | GeneralSearchFormat keywordFunc, search when search.Length > 0 -> search |> keywordFunc |> runApiSearch
+        | SpecificSearchFormat keywordFunc, UserRepoFormat search
+        | UserRepoFormat search, SpecificSearchFormat keywordFunc -> search |> keywordFunc |> runApiSearch
+        | UserRepoFormat (u, r), IssueFormat i -> runApiSearch (FindIssue(u, r, i))
+        | search, "" -> SuggestQuery(SearchRepos search)
+        | _ -> SuggestQuery DefaultSuggestion
```

Comment:
Here you could have kept the `parseQuery` function as-is except for the first two match cases. 

On the matching side, you can use the `::` operator for these two cases to deconstruct the input into the head (either "repos" or "users") and the tail (list of space separated query words).

On the right side, you can re-join the tail into a single string `String.concat " "`

Reply:
I check Flow's code. It may be better to let Flow handle the split because the Term field is simply split from user search, so if the actionkeyword is *, it would be possible to skip a search term unexpected.

Reply:
Good point, then perhaps something like

```fsharp
|> List.skip (if query.ActionKeyword = query.GlobalPluginWildcardSign then 0 else 1)
```

in `QueryAsync` ?

Reply:
(though, realistically, who in their right mind would use * for the github plugin 😛)

Reply:
I extract the list from Query.FirstSearch and Query.SecondToEndSearch. What do you think about that👀

Reply:
I prefer the old way for a few reasons...

The way you have it, `parseQuery` will always get a list of exactly 2 elements, which is misleading and defeats the purpose of passing a list, and also does not allow for pattern matching to be fully utilised on the typed query. You also have to do awkward changes to the tests like you did in 352e26a4caf7e1629dd18d4e960519c69a19dcec to get them passing...

I also think it's much cleaner to have all query cases in a single function instead of splitting them up, I find them easier to read/reason about that way.

Another minor point is that I'd like to keep the differences between Wox and Flow plugins to the minimum.

Anyway, I haven't tested the following but I'm thinking of something like this:

```fsharp
let parseQuery = function
    | "repos" :: search                       -> runApiSearch (FindRepos (String.concat " " search))
    | "users" :: search                       -> runApiSearch (FindUsers (String.concat " " search))
    | [ "issues"; UserRepoFormat search ]     -> runApiSearch (FindIssues search)
    | [ "pr";     UserRepoFormat search ]     -> runApiSearch (FindPRs search)
    | [ "pull";   UserRepoFormat search ]     -> runApiSearch (FindPRs search)
    | [ "repo";   UserRepoFormat search ]     -> runApiSearch (FindRepo search)
    | [ UserRepoFormat search           ]     -> runApiSearch (FindRepo search)
    | [ UserRepoFormat search; "issues" ]     -> runApiSearch (FindIssues search)
    | [ UserRepoFormat search; "pr"     ]     -> runApiSearch (FindPRs search)
    | [ UserRepoFormat search; "pull"   ]     -> runApiSearch (FindPRs search)
    | [ UserRepoFormat (u,r); IssueFormat i ] -> runApiSearch (FindIssue (u,r,i))
    | [ UserReposFormat user ]                -> runApiSearch (FindUserRepos user)
    | [ search ]                              -> SuggestQuery (SearchRepos search)
    | _                                       -> SuggestQuery DefaultSuggestion
```

and then the `QueryAsync` overload can look like this:

```fsharp
member this.QueryAsync(query, token: CancellationToken) =
    let ghSearch = async {
        let! results =
            query.Terms
            |> List.ofArray
            |> List.skip (if query.ActionKeyword = query.GlobalPluginWildcardSign then 0 else 1)
            |> this.ProcessQuery

        return results
                |> List.map (fun r -> Result( Title = r.title, SubTitle = r.subtitle, IcoPath = "icon.png", Action = fun x -> r.action x ))
                |> List<Result>
    }
    Async.StartImmediateAsTask(ghSearch, token)
```

Reply:
Got you😉